text/0000-stabilize-marker-freeze.md

```diff
@@ -0,0 +1,66 @@
+- Feature Name: `stabilize_marker_freeze`
+- Start Date: 2024-05-10
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Stabilize `core::marker::Freeze` in trait bounds.
+
+# Motivation
+[motivation]: #motivation
+
+In some contexts, guaranteeing that a value cannot be modified is a requirement to a construct behaving properly (`const` references, keys of a structured map...).
```

Comment:
It's not clear to me how this helps for "keys of a structured map". Consider that `Box<T>: Freeze` for all `T` -- there can be interior mutability behind pointer indirection even with `Freeze`.

`Freeze` is entirely a by-value, "shallow" property.

Reply:
damn, that makes it way less useful for other purposes than static promotion then :(

Reply:
As an aside, could changing this shallowness property be an interesting direction for this RFC, or is the current shallowness important in other aspects at the moment?

Reply:
It's crucial for it to remain shallow. That's the entire point the trait was introduced for: determine whether shared references permit mutation or not. In an `&Box<Cell<T>>`, the data behind the first pointer is fully immutable and hence the `Box` should be `Freeze`.